src/ml/neural_net/model_spec.hpp

```diff
@@ -395,6 +395,13 @@ class model_spec {
    */
   void add_preprocessing(const std::string& feature_name,
                          const float image_scale);
+  void change_preprocessing_name(const std::string& old_name,
+                                 const std::string& new_name);
+
+  /** Set input name for given layer
+   */
+  void set_layer_input(const std::string& layer_name,
+                       const std::string& input_name, const size_t index);
```

Comment:
It feels strange to me that we should need these.... I'll follow up below where they are used, but it seems weird to need to modify the spec after we've built it. Ideally we can build it correctly the first time, as necessary

Reply:
It is weird, the reason I do here is in the `init_model()`, the `nn_spec_` is read from a pretrain model file:
```
std::unique_ptr<model_spec> nn_spec(new model_spec(pretrained_mlmodel_path));
```
which fixed the input name to "image", which might not be the feature column name the user gives.
Currently I don't have a better solutions for this weird usage.
But yes for the `import_from_custom()` I can directly construct a correct one.

Reply:
Let's set the names during export, not in the main toolkit. That code can use the underlying proto code if needed, not the `model_spec` abstraction.

We need to keep this abstraction clean, so it's more obvious how to use it when defining a model from scratch